src/main/java/org/jboss/msc/service/ServiceContainerImpl.java

```diff
@@ -111,23 +109,12 @@ final class ServiceContainerImpl extends ServiceTargetImpl implements ServiceCon
     private final List<TerminateListener> terminateListeners = new ArrayList<>(1);
 
     private static final class ShutdownHookThread extends Thread {
-        final Reference<ServiceContainer, Void> containerRef;
+        final Reference<ServiceContainer> containerRef;
 
         private ShutdownHookThread(final ServiceContainer container) {
             setName(container.getName() + " MSC Shutdown Thread");
             setDaemon(false);
-            containerRef = new WeakReference<>(container, null, new Reaper<ServiceContainer, Void>() {
-                public void reap(final Reference<ServiceContainer, Void> containerRef) {
-                    final ServiceContainer container = containerRef.get();
-                    if (container == null) return;
-                    container.shutdown();
-                    try {
-                        container.awaitTermination();
-                    } catch (InterruptedException ie) {
-                        // ignored
-                    }
-                }
-            });
+            containerRef = new WeakReference<>(container);
```

Comment:
The container will never be shut down in this case when it becomes weakly-reachable. What about switching to `smallrye-common-ref` instead, which has this exact same API (only not deprecated)?

Reply:
Ah I missed that comment, thanks. The change does make more sense now, however it is important to note that finalization is deprecated for removal in the JDK so we do need a different solution.

It may actually not be possible to solve this problem because the graph of objects will always contain the container itself. We might have to remove this entire leak-prevention mechanism and make wildfly-core be in charge of it (using phantom references or something).